File: leetcode/c++/0128-longest-consecutive-sequence.cc
// https://leetcode.com/problems/longest-consecutive-sequence/
// Hard

#include<vector>
#include<unordered_map>
using namespace std;

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        if (nums.empty()) return 0;
        unordered_map<int, int> m;
        for (int &n: nums) {
            m[n] = 1;
        }
        int maxLen = 1;
        int tempN = 0;
        for (int n: nums) {
            if (m.find(n-1) != m.end()) continue;
            tempN = n;
            while (m.find(n+1) != m.end()) {
                n++;
            }
            maxLen = max(maxLen, n-tempN+1);
        }
        // for (pair<int, int> p: m) {
        //     cout << p.first << ' ' << p.second << endl;
        // }
        // cout << endl;
        return maxLen;
    }
};

File: leetcode/python/0128-longest-consecutive-sequence.py
# https://leetcode.com/problems/longest-consecutive-sequence/
# Medium

# Create set containing all the elements
# For all n in nums, go through all n,n+1,n+2,...,n+k if n-1 does not exist in set
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        s = set(nums)
        maxLen = 0
        for n in nums:
            if n-1 not in s:
                l = n
                while l in s:
                    l += 1
                maxLen = max(maxLen, l-n)
        return maxLen